artiq/examples/kasli_shuttler/repository/shuttler.py

```diff
@@ -0,0 +1,77 @@
+from artiq.experiment import *
+
+
+class Shuttler(EnvExperiment):
+    def build(self):
+        self.setattr_device("core")
+        self.leds = [ self.get_device("efc_led{}".format(i)) for i in range(2) ]
+
+        self.setattr_device("pdq_config")
+        self.setattr_device("pdq_trigger")
+        self.pdq_volt = [ self.get_device("pdq{}_volt".format(i)) for i in range(16) ]
+        self.pdq_dds = [ self.get_device("pdq{}_dds".format(i)) for i in range(16) ]
+
+    @kernel
+    def run(self):
+        self.init()
+
+    @kernel
+    def led(self):
+        self.core.break_realtime()
+        for i in range(2):
+            for j in range(3):
+                self.leds[i].pulse(.1*s)
+                delay(.1*s)
+
+    @kernel
+    def pdq_volt_mu(self, volt):
+        return round(2 ** 13 / 5 * volt)
+
+    @kernel
+    def pdq_freq_mu(self, freq_mhz):
+        return round(float(2) ** 32 / 125 * freq_mhz)
```

Comment:
These functions seem useful beyond the example code - should they be a part of the coredevice driver?

Reply:
Speaking of this, are we truncating 2 MSBs from Shuttler core?

Reply:
> Speaking of this, are we truncating 2 MSBs from Shuttler core? The `2 ** 13` factor seems to imply this.

Yes. 2 MSBs are truncated from Shuttler core as AD9117 is a 14bit DAC.

Reply:
If you truncate the MSBs from Shuttler core, does it make it hard to detect when the output waveform is digitally clipping? Would it not be better to truncate the 2 LSBs?

Reply:
In addition, if/when we get around to implementing sigma-delta modulation to increase bit depth, we will want to have more than 14 bits. This works much more naturally if we are truncating the 2 LSBs for 14 bits, and then keeping one or both if we go to 15 or 16 bits using sigma-delta.

Reply:
@sbourdeauducq the PDQ core should run at a higher resolution than the output, otherwise the truncation errors during the waveform computation from e.g. summing offset with DDS tones will compound and degrade the quality of the output signal. Also if we plan for future sigma-delta we will need even more bits in the PDQ core.